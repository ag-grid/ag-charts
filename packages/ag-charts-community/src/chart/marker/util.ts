import { Circle } from './circle';
import { Cross } from './cross';
import { Diamond } from './diamond';
import { Heart } from './heart';
import type { Marker } from './marker';
import { Plus } from './plus';
import { Square } from './square';
import { Triangle } from './triangle';

type MarkerConstructor = new () => Marker;
type MarkerSupportedShapes = 'circle' | 'cross' | 'diamond' | 'heart' | 'plus' | 'square' | 'triangle';
export type MarkerShape = MarkerConstructor | MarkerSupportedShapes;

const MARKER_SHAPES: { [K in MarkerSupportedShapes]: MarkerConstructor } = {
    circle: Circle,
    cross: Cross,
    diamond: Diamond,
    heart: Heart,
    plus: Plus,
    square: Square,
    triangle: Triangle,
};

const MARKER_SUPPORTED_SHAPES = Object.keys(MARKER_SHAPES);

export function isMarkerShape(shape: MarkerShape | string): shape is MarkerSupportedShapes {
    return typeof shape === 'string' && MARKER_SUPPORTED_SHAPES.includes(shape);
}

// This function is in its own file because putting it into SeriesMarker makes the Legend
// suddenly aware of the series (it's an agnostic component), and putting it into Marker
// introduces circular dependencies.
export function getMarker(shape: MarkerShape | string = Square): MarkerConstructor {
    if (isMarkerShape(shape)) {
        return MARKER_SHAPES[shape];
    }
    if (typeof shape === 'function') {
        return shape;
    }
    return Square;
}
