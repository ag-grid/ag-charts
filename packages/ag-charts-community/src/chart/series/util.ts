import type { ChartAxis } from '../chartAxis';
import type { ProcessedDataDef } from '../data/dataModel';

export function convertValuesToScaleByDefs<T extends string>({
    defs,
    values,
    xAxis,
    yAxis,
}: {
    defs: [string, ProcessedDataDef[]][];
    values: Record<T, unknown>;
    xAxis: ChartAxis | undefined;
    yAxis: ChartAxis | undefined;
}): Record<T, number> {
    if (!(xAxis && yAxis)) {
        throw new Error('Axes must be defined');
    }
    const result: Record<string, number> = {};
    for (const [searchId, [{ def }]] of defs) {
        if (Object.hasOwn(values, searchId)) {
            const { scale } = def.type === 'key' ? xAxis : yAxis;
            result[searchId] = Math.round(scale.convert((values as any)[searchId]));
        }
    }
    return result;
}
