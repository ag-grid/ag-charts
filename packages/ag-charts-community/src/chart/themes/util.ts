import type { AgCartesianChartOptions } from 'ag-charts-types';

type CartesianAxis = Exclude<AgCartesianChartOptions['axes'], undefined>[0];

export function swapAxisCondition(axes: [CartesianAxis, CartesianAxis], swap: (series: any) => boolean) {
    return (series: any) => {
        if (swap(series)) {
            return [
                { ...axes[0], position: axes[1].position },
                { ...axes[1], position: axes[0].position },
            ];
        } else {
            return axes;
        }
    };
}
