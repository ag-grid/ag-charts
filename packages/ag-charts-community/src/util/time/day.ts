import { CountableTimeInterval } from './interval';
import { durationDay } from './duration';

function encode(date: Date) {
    const tzOffsetMs = date.getTimezoneOffset() * 60_000;

    return Math.floor((date.getTime() - tzOffsetMs) / durationDay);
}

function decode(encoded: number) {
    const d = new Date(0);
    const startTzOffsetMs = d.getTimezoneOffset() * 60_000;
    d.setTime(d.getTime() + startTzOffsetMs);
    d.setDate(d.getDate() + encoded);

    return d;
}

export const day = new CountableTimeInterval(encode, decode);
export default day;
