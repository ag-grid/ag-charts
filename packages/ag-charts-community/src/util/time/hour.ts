import { durationHour, durationMinute } from './duration';
import { CountableTimeInterval } from './interval';

const offset = new Date().getTimezoneOffset() * durationMinute;

function encode(date: Date) {
    return Math.floor((date.getTime() - offset) / durationHour);
}

function decode(encoded: number) {
    return new Date(offset + encoded * durationHour);
}

export const hour = new CountableTimeInterval(encode, decode);
export default hour;
