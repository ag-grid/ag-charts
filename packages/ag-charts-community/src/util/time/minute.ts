import { CountableTimeInterval } from './interval';
import { durationMinute } from './duration';

const offset = new Date().getTimezoneOffset() * durationMinute;

function encode(date: Date) {
    return Math.floor((date.getTime() - offset) / durationMinute);
}

function decode(encoded: number) {
    return new Date(offset + encoded * durationMinute);
}

export const minute = new CountableTimeInterval(encode, decode);
export default minute;
