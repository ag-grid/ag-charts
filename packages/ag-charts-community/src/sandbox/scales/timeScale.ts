import { convertDomain } from '../util/domain.util';
import { ContinuousScale } from './continuousScale';

export class TimeScale extends ContinuousScale<Date | number, number> {
    constructor(domain: (Date | number)[], range: number[]) {
        super(domain.map(Number), range);
    }

    convert(value: Date | number, clamp?: boolean) {
        return convertDomain(Number(value), this.domain as number[], this.range, clamp);
    }

    invert(value: number, clamp?: boolean) {
        return new Date(convertDomain(value, this.range, this.domain as number[], clamp));
    }
}
