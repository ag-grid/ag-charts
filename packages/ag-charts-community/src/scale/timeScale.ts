import { findMinMax } from '../util/number';
import { TickIntervals, getTickInterval, isDenseInterval } from '../util/ticks';
import { TimeInterval } from '../util/time/interval';
import { buildFormatter } from '../util/timeFormat';
import { dateToNumber, defaultTimeTickFormat } from '../util/timeFormatDefaults';
import { ContinuousScale } from './continuousScale';

export class TimeScale extends ContinuousScale<Date, TimeInterval | number> {
    readonly type = 'time';

    public constructor() {
        super([], [0, 1]);
    }

    toDomain(d: number): Date {
        return new Date(d);
    }

    override convert(value: Date, clamp?: boolean): number {
        return super.convert(new Date(value), clamp);
    }

    override invert(value: number): Date {
        return new Date(super.invert(value));
    }

    /**
     * Returns uniformly-spaced dates that represent the scale's domain.
     */
    ticks(): Date[] {
        if (!this.domain || this.domain.length < 2) {
            return [];
        }
        this.refresh();

        const { interval, nice, tickCount, minTickCount, maxTickCount } = this;
        const [start, stop] = findMinMax(this.getDomain().map(dateToNumber));

        if (interval != null) {
            return (
                TimeScale.getTicksForInterval({ start, stop, interval, availableRange: this.getPixelRange() }) ??
                TimeScale.getDefaultTicks({ start, stop, tickCount, minTickCount, maxTickCount })
            );
        } else if (nice && tickCount === 2) {
            return this.niceDomain;
        } else if (nice && tickCount === 1) {
            return this.niceDomain.slice(0, 1);
        }
        return TimeScale.getDefaultTicks({ start, stop, tickCount, minTickCount, maxTickCount });
    }

    static getDefaultTicks({
        start,
        stop,
        tickCount,
        minTickCount,
        maxTickCount,
    }: {
        start: number;
        stop: number;
        tickCount: number;
        minTickCount: number;
        maxTickCount: number;
    }) {
        const t = getTickInterval(start, stop, tickCount, minTickCount, maxTickCount);
        return t ? t.range(new Date(start), new Date(stop)) : []; // inclusive stop
    }

    static getTicksForInterval({
        start,
        stop,
        interval,
        availableRange,
    }: {
        start: number;
        stop: number;
        interval: number | TimeInterval;
        availableRange: number;
    }): Date[] | undefined {
        if (!interval) {
            return [];
        }

        if (interval instanceof TimeInterval) {
            const ticks = interval.range(new Date(start), new Date(stop));
            if (isDenseInterval(ticks.length, availableRange)) {
                return;
            }

            return ticks;
        }

        const absInterval = Math.abs(interval);

        if (isDenseInterval((stop - start) / absInterval, availableRange)) return;

        const timeInterval = TickIntervals.findLast((tickInterval) => absInterval % tickInterval.duration === 0);

        if (timeInterval) {
            const i = timeInterval.timeInterval.every(absInterval / (timeInterval.duration / timeInterval.step));
            return i.range(new Date(start), new Date(stop));
        }

        let date = new Date(start);
        const stopDate = new Date(stop);
        const ticks = [];
        while (date <= stopDate) {
            ticks.push(date);
            date = new Date(date);
            date.setMilliseconds(date.getMilliseconds() + absInterval);
        }

        return ticks;
    }

    /**
     * Returns a time format function suitable for displaying tick values.
     *
     * @param ticks Optional array of tick values for custom formatting.
     * @param domain Optional array representing the [min, max] values of the time axis.
     * @param specifier Optional format specifier string for custom date formatting (e.g., `%Y`, `%m`, `%d`).
     * @param formatOffset Optional number for applying an offset to the format (e.g., timezone shifts).
     * @returns A function that formats a `Date` object into a string based on the provided specifier or default format.
     */
    tickFormat({
        ticks,
        domain,
        specifier,
        formatOffset,
    }: {
        ticks?: any[];
        domain?: any[];
        specifier?: string;
        formatOffset?: number;
    }): (date: Date) => string {
        return specifier == null ? defaultTimeTickFormat(ticks, domain, formatOffset) : buildFormatter(specifier);
    }

    update() {
        if (!this.domain || this.domain.length < 2) {
            return;
        }
        if (this.nice) {
            this.updateNiceDomain();
        }
    }

    /**
     * Extends the domain so that it starts and ends on nice round values.
     * This method typically modifies the scaleâ€™s domain, and may only extend the bounds to the nearest round value.
     */
    protected updateNiceDomain(): void {
        const maxAttempts = 4;
        let [d0, d1] = this.domain;
        for (let i = 0; i < maxAttempts; i++) {
            this.updateNiceDomainIteration(d0, d1);
            const [n0, n1] = this.niceDomain;
            if (dateToNumber(d0) === dateToNumber(n0) && dateToNumber(d1) === dateToNumber(n1)) {
                break;
            }
            d0 = n0;
            d1 = n1;
        }
    }

    protected updateNiceDomainIteration(d0: Date, d1: Date) {
        const start = Math.min(dateToNumber(d0), dateToNumber(d1));
        const stop = Math.max(dateToNumber(d0), dateToNumber(d1));

        const isReversed = d0 > d1;

        const { interval } = this;
        let i;

        if (interval instanceof TimeInterval) {
            i = interval;
        } else {
            const tickCount = typeof interval === 'number' ? (stop - start) / Math.max(interval, 1) : this.tickCount;
            i = getTickInterval(start, stop, tickCount, this.minTickCount, this.maxTickCount);
        }

        if (i) {
            const intervalRange = i.range(new Date(start), new Date(stop), true);
            const domain = isReversed ? [...intervalRange].reverse() : intervalRange;
            const n0 = domain[0];
            const n1 = domain.at(-1)!;
            this.niceDomain = [n0, n1];
        }
    }
}
