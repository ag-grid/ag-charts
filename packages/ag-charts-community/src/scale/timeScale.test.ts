import { describe, expect, it } from '@jest/globals';

import day from '../util/time/day';
import {
    durationDay,
    durationHour,
    durationMinute,
    durationMonth,
    durationWeek,
    durationYear,
} from '../util/time/duration';
import hour from '../util/time/hour';
import minute from '../util/time/minute';
import month from '../util/time/month';
import year from '../util/time/year';
import { calculateDefaultTimeTickFormat } from '../util/timeFormatDefaults';
import { TimeScale } from './timeScale';

describe('TimeScale', () => {
    it('should create nice domain', () => {
        const scale = new TimeScale();
        scale.domain = [new Date(new Date(2022, 1, 13)), new Date(new Date(2022, 10, 30))];
        scale.nice = true;
        scale.update();
        expect(scale.getDomain()).toEqual([new Date(2022, 1, 1), new Date(2022, 11, 1)]);
    });

    it('should create nice ticks', () => {
        const scale = new TimeScale();
        scale.domain = [new Date(2022, 1, 13), new Date(2022, 10, 30)];
        scale.nice = true;
        scale.tickCount = 10;
        expect(scale.ticks()).toEqual([
            new Date(2022, 1, 1),
            new Date(2022, 2, 1),
            new Date(2022, 3, 1),
            new Date(2022, 4, 1),
            new Date(2022, 5, 1),
            new Date(2022, 6, 1),
            new Date(2022, 7, 1),
            new Date(2022, 8, 1),
            new Date(2022, 9, 1),
            new Date(2022, 10, 1),
            new Date(2022, 11, 1),
        ]);
    });

    describe('should create ticks with configured', () => {
        describe(`milliseconds interval`, () => {
            const scale = new TimeScale();
            scale.nice = true;

            const MILLISECONDS_INTERVALS = [
                {
                    name: 'every minute',
                    interval: durationMinute,
                    domain: [new Date(2022, 0, 1, 12), new Date(2022, 0, 1, 13)],
                },
                {
                    name: 'every hour',
                    interval: durationHour,
                    domain: [new Date(2022, 0, 1, 9), new Date(2022, 0, 1, 17)],
                },
                {
                    name: 'every half day',
                    interval: durationDay / 2,
                    domain: [new Date(2022, 0, 1), new Date(2022, 0, 15)],
                },
                {
                    name: 'every 2 days',
                    interval: durationDay * 2,
                    domain: [new Date(2022, 0, 1), new Date(2022, 0, 21)],
                },
                {
                    name: 'every 3 weeks',
                    interval: durationWeek * 3,
                    domain: [new Date(2022, 0, 1), new Date(2022, 6, 1)],
                },
                {
                    name: 'every month',
                    interval: durationMonth,
                    domain: [new Date(2022, 0, 1), new Date(2023, 0, 1)],
                },
                {
                    name: 'every two months',
                    interval: durationMonth * 2,
                    domain: [new Date(2022, 0, 1), new Date(2023, 0, 1)],
                },
                {
                    name: 'every year',
                    interval: durationYear,
                    domain: [new Date(2021, 0, 1), new Date(2023, 0, 1)],
                },
                {
                    name: 'every year, negative milliseconds',
                    interval: -durationYear,
                    domain: [new Date(2021, 0, 1), new Date(2023, 0, 1)],
                },
            ];

            it.each(MILLISECONDS_INTERVALS)(`for $name case`, ({ interval, domain }) => {
                const scale2 = new TimeScale();

                scale2.range = [0, 600];
                scale2.domain = domain;
                scale2.interval = interval;

                expect(scale2.ticks()).toMatchSnapshot();
            });
        });

        describe(`time interval`, () => {
            const scale = new TimeScale();
            scale.nice = true;

            const TIME_INTERVALS = [
                {
                    name: 'every minute',
                    interval: minute,
                    domain: [new Date(2022, 0, 1, 12), new Date(2022, 0, 1, 13)],
                },
                {
                    name: 'every hour',
                    interval: hour,
                    domain: [new Date(2022, 0, 1, 9), new Date(2022, 0, 1, 17)],
                },
                {
                    name: 'every day',
                    interval: day,
                    domain: [new Date(2022, 0, 1), new Date(2022, 0, 15)],
                },
                {
                    name: 'every 3 days',
                    interval: day.every(3),
                    domain: [new Date(2022, 0, 1), new Date(2022, 0, 21)],
                },
                {
                    name: 'every month',
                    interval: month,
                    domain: [new Date(2022, 0, 1), new Date(2022, 6, 1)],
                },
                {
                    name: 'every two months',
                    interval: month.every(2),
                    domain: [new Date(2022, 0, 1), new Date(2023, 0, 1)],
                },
                {
                    name: 'every 6 months',
                    interval: month.every(6),
                    domain: [new Date(2021, 0, 1), new Date(2023, 0, 1)],
                },
                {
                    name: 'every year',
                    interval: year,
                    domain: [new Date(2021, 0, 1), new Date(2023, 0, 1)],
                },
            ];

            it.each(TIME_INTERVALS)(`for $name case`, ({ interval, domain }) => {
                const scale2 = new TimeScale();

                scale2.range = [0, 600];
                scale2.domain = domain;
                scale2.interval = interval;

                expect(scale2.ticks()).toMatchSnapshot();
            });
        });
    });

    describe('#calculateDefaultTickFormat', () => {
        const TEST_CASES: { name: string; ticks: Date[]; expectedFormat: string }[] = [
            {
                name: 'several seconds',
                ticks: [
                    new Date(2023, 0, 1, 13, 25, 30),
                    new Date(2023, 0, 1, 13, 25, 35),
                    new Date(2023, 0, 1, 13, 25, 40),
                    new Date(2023, 0, 1, 13, 25, 45),
                ],
                expectedFormat: ':%S', //  second
            },
            {
                name: 'several minutes',
                ticks: [
                    new Date(2023, 0, 1, 13, 25),
                    new Date(2023, 0, 1, 13, 26),
                    new Date(2023, 0, 1, 13, 27),
                    new Date(2023, 0, 1, 13, 28),
                ],
                expectedFormat: '%I:%M', // hour:minute
            },
            {
                name: 'several hours',
                ticks: [
                    new Date(2023, 0, 1, 13),
                    new Date(2023, 0, 1, 15),
                    new Date(2023, 0, 1, 17),
                    new Date(2023, 0, 1, 19),
                ],
                expectedFormat: '%I %p', // hour
            },
            {
                name: 'several days',
                ticks: [new Date(2022, 0, 1), new Date(2022, 0, 2), new Date(2022, 0, 3)],
                expectedFormat: '%a', // week day
            },
            {
                name: 'several months',
                ticks: [new Date(2022, 0, 1), new Date(2022, 1, 1), new Date(2022, 2, 1)],
                expectedFormat: '%B', // full month
            },
            {
                name: 'several years',
                ticks: [new Date(2021, 0, 1), new Date(2022, 0, 1), new Date(2023, 0, 1)],
                expectedFormat: '%Y', // year
            },
            {
                name: 'several seconds spanning minutes',
                ticks: [
                    new Date(2023, 0, 1, 13, 25, 0),
                    new Date(2023, 0, 1, 13, 25, 30),
                    new Date(2023, 0, 1, 13, 26, 0),
                    new Date(2023, 0, 1, 13, 26, 30),
                    new Date(2023, 0, 1, 13, 27, 0),
                ],
                expectedFormat: '%I:%M:%S', // hour:minute:second
            },
            {
                name: 'several minutes spanning hours',
                ticks: [new Date(2023, 0, 1, 13, 25), new Date(2023, 0, 1, 15, 25), new Date(2023, 0, 1, 17, 25)],
                expectedFormat: '%I %p%I:%M',
            },
            {
                name: 'several hours spanning days',
                ticks: [
                    new Date(2023, 0, 1, 12),
                    new Date(2023, 0, 2, 0),
                    new Date(2023, 0, 2, 12),
                    new Date(2023, 0, 3, 0),
                ],
                expectedFormat: '%I %p %a', // hour + week day
            },
            {
                name: 'several hours spanning weeks',
                ticks: [
                    new Date(2023, 0, 1, 12),
                    new Date(2023, 0, 2, 0),
                    new Date(2023, 0, 2, 12),
                    new Date(2023, 0, 14, 0),
                ],
                expectedFormat: '%I %p %b %d', // hour + short month + date
            },
            {
                name: 'several days spanning months',
                ticks: [new Date(2022, 0, 30), new Date(2022, 0, 31), new Date(2022, 2, 1), new Date(2022, 2, 2)],
                expectedFormat: '%b %d', // short month + date
            },
            {
                name: 'several days across years',
                ticks: [new Date(2022, 11, 28), new Date(2022, 11, 30), new Date(2023, 0, 2), new Date(2023, 0, 4)],
                expectedFormat: '%b %d %Y', // short month + date + year
            },
            {
                name: 'several months spanning years',
                ticks: [
                    new Date(2022, 0, 1),
                    new Date(2022, 3, 1),
                    new Date(2022, 6, 1),
                    new Date(2022, 9, 1),
                    new Date(2023, 0, 1),
                ],
                expectedFormat: '%B %Y', // full month + year
            },
            {
                name: 'several months across years',
                ticks: [new Date(2022, 10, 1), new Date(2022, 11, 1), new Date(2023, 0, 1)],
                expectedFormat: '%B %Y', // full month + year
            },
        ];

        it.each(TEST_CASES)(`for $name case`, ({ ticks, expectedFormat }) => {
            const scale = new TimeScale();
            scale.domain = [ticks[0], ticks[ticks.length - 1]];

            expect(calculateDefaultTimeTickFormat(ticks, scale.buildFormatString)).toEqual(expectedFormat);
        });
    });
});
