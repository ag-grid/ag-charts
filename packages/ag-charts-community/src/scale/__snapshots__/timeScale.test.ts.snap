// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeScale should create ticks with configured milliseconds interval for every 2 days case 1`] = `
Array [
  2022-01-01T00:00:00.000Z,
  2022-01-03T00:00:00.000Z,
  2022-01-05T00:00:00.000Z,
  2022-01-07T00:00:00.000Z,
  2022-01-09T00:00:00.000Z,
  2022-01-11T00:00:00.000Z,
  2022-01-13T00:00:00.000Z,
  2022-01-15T00:00:00.000Z,
  2022-01-17T00:00:00.000Z,
  2022-01-19T00:00:00.000Z,
  2022-01-21T00:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured milliseconds interval for every 3 weeks case 1`] = `
Array [
  2022-01-16T00:00:00.000Z,
  2022-02-06T00:00:00.000Z,
  2022-02-27T00:00:00.000Z,
  2022-03-20T00:00:00.000Z,
  2022-04-09T23:00:00.000Z,
  2022-04-30T23:00:00.000Z,
  2022-05-21T23:00:00.000Z,
  2022-06-11T23:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured milliseconds interval for every half day case 1`] = `
Array [
  2022-01-01T00:00:00.000Z,
  2022-01-01T12:00:00.000Z,
  2022-01-02T00:00:00.000Z,
  2022-01-02T12:00:00.000Z,
  2022-01-03T00:00:00.000Z,
  2022-01-03T12:00:00.000Z,
  2022-01-04T00:00:00.000Z,
  2022-01-04T12:00:00.000Z,
  2022-01-05T00:00:00.000Z,
  2022-01-05T12:00:00.000Z,
  2022-01-06T00:00:00.000Z,
  2022-01-06T12:00:00.000Z,
  2022-01-07T00:00:00.000Z,
  2022-01-07T12:00:00.000Z,
  2022-01-08T00:00:00.000Z,
  2022-01-08T12:00:00.000Z,
  2022-01-09T00:00:00.000Z,
  2022-01-09T12:00:00.000Z,
  2022-01-10T00:00:00.000Z,
  2022-01-10T12:00:00.000Z,
  2022-01-11T00:00:00.000Z,
  2022-01-11T12:00:00.000Z,
  2022-01-12T00:00:00.000Z,
  2022-01-12T12:00:00.000Z,
  2022-01-13T00:00:00.000Z,
  2022-01-13T12:00:00.000Z,
  2022-01-14T00:00:00.000Z,
  2022-01-14T12:00:00.000Z,
  2022-01-15T00:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured milliseconds interval for every hour case 1`] = `
Array [
  2022-01-01T09:00:00.000Z,
  2022-01-01T10:00:00.000Z,
  2022-01-01T11:00:00.000Z,
  2022-01-01T12:00:00.000Z,
  2022-01-01T13:00:00.000Z,
  2022-01-01T14:00:00.000Z,
  2022-01-01T15:00:00.000Z,
  2022-01-01T16:00:00.000Z,
  2022-01-01T17:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured milliseconds interval for every minute case 1`] = `
Array [
  2022-01-01T12:00:00.000Z,
  2022-01-01T12:01:00.000Z,
  2022-01-01T12:02:00.000Z,
  2022-01-01T12:03:00.000Z,
  2022-01-01T12:04:00.000Z,
  2022-01-01T12:05:00.000Z,
  2022-01-01T12:06:00.000Z,
  2022-01-01T12:07:00.000Z,
  2022-01-01T12:08:00.000Z,
  2022-01-01T12:09:00.000Z,
  2022-01-01T12:10:00.000Z,
  2022-01-01T12:11:00.000Z,
  2022-01-01T12:12:00.000Z,
  2022-01-01T12:13:00.000Z,
  2022-01-01T12:14:00.000Z,
  2022-01-01T12:15:00.000Z,
  2022-01-01T12:16:00.000Z,
  2022-01-01T12:17:00.000Z,
  2022-01-01T12:18:00.000Z,
  2022-01-01T12:19:00.000Z,
  2022-01-01T12:20:00.000Z,
  2022-01-01T12:21:00.000Z,
  2022-01-01T12:22:00.000Z,
  2022-01-01T12:23:00.000Z,
  2022-01-01T12:24:00.000Z,
  2022-01-01T12:25:00.000Z,
  2022-01-01T12:26:00.000Z,
  2022-01-01T12:27:00.000Z,
  2022-01-01T12:28:00.000Z,
  2022-01-01T12:29:00.000Z,
  2022-01-01T12:30:00.000Z,
  2022-01-01T12:31:00.000Z,
  2022-01-01T12:32:00.000Z,
  2022-01-01T12:33:00.000Z,
  2022-01-01T12:34:00.000Z,
  2022-01-01T12:35:00.000Z,
  2022-01-01T12:36:00.000Z,
  2022-01-01T12:37:00.000Z,
  2022-01-01T12:38:00.000Z,
  2022-01-01T12:39:00.000Z,
  2022-01-01T12:40:00.000Z,
  2022-01-01T12:41:00.000Z,
  2022-01-01T12:42:00.000Z,
  2022-01-01T12:43:00.000Z,
  2022-01-01T12:44:00.000Z,
  2022-01-01T12:45:00.000Z,
  2022-01-01T12:46:00.000Z,
  2022-01-01T12:47:00.000Z,
  2022-01-01T12:48:00.000Z,
  2022-01-01T12:49:00.000Z,
  2022-01-01T12:50:00.000Z,
  2022-01-01T12:51:00.000Z,
  2022-01-01T12:52:00.000Z,
  2022-01-01T12:53:00.000Z,
  2022-01-01T12:54:00.000Z,
  2022-01-01T12:55:00.000Z,
  2022-01-01T12:56:00.000Z,
  2022-01-01T12:57:00.000Z,
  2022-01-01T12:58:00.000Z,
  2022-01-01T12:59:00.000Z,
  2022-01-01T13:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured milliseconds interval for every month case 1`] = `
Array [
  2022-01-01T00:00:00.000Z,
  2022-02-01T00:00:00.000Z,
  2022-03-01T00:00:00.000Z,
  2022-03-31T23:00:00.000Z,
  2022-04-30T23:00:00.000Z,
  2022-05-31T23:00:00.000Z,
  2022-06-30T23:00:00.000Z,
  2022-07-31T23:00:00.000Z,
  2022-08-31T23:00:00.000Z,
  2022-09-30T23:00:00.000Z,
  2022-11-01T00:00:00.000Z,
  2022-12-01T00:00:00.000Z,
  2023-01-01T00:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured milliseconds interval for every two months case 1`] = `
Array [
  2022-01-01T00:00:00.000Z,
  2022-03-01T00:00:00.000Z,
  2022-04-30T23:00:00.000Z,
  2022-06-30T23:00:00.000Z,
  2022-08-31T23:00:00.000Z,
  2022-11-01T00:00:00.000Z,
  2023-01-01T00:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured milliseconds interval for every year case 1`] = `
Array [
  2021-01-01T00:00:00.000Z,
  2022-01-01T00:00:00.000Z,
  2023-01-01T00:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured milliseconds interval for every year, negative milliseconds case 1`] = `
Array [
  2021-01-01T00:00:00.000Z,
  2022-01-01T00:00:00.000Z,
  2023-01-01T00:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured time interval for every 3 days case 1`] = `
Array [
  2022-01-01T00:00:00.000Z,
  2022-01-04T00:00:00.000Z,
  2022-01-07T00:00:00.000Z,
  2022-01-10T00:00:00.000Z,
  2022-01-13T00:00:00.000Z,
  2022-01-16T00:00:00.000Z,
  2022-01-19T00:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured time interval for every 6 months case 1`] = `
Array [
  2021-01-01T00:00:00.000Z,
  2021-06-30T23:00:00.000Z,
  2022-01-01T00:00:00.000Z,
  2022-06-30T23:00:00.000Z,
  2023-01-01T00:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured time interval for every day case 1`] = `
Array [
  2022-01-01T00:00:00.000Z,
  2022-01-02T00:00:00.000Z,
  2022-01-03T00:00:00.000Z,
  2022-01-04T00:00:00.000Z,
  2022-01-05T00:00:00.000Z,
  2022-01-06T00:00:00.000Z,
  2022-01-07T00:00:00.000Z,
  2022-01-08T00:00:00.000Z,
  2022-01-09T00:00:00.000Z,
  2022-01-10T00:00:00.000Z,
  2022-01-11T00:00:00.000Z,
  2022-01-12T00:00:00.000Z,
  2022-01-13T00:00:00.000Z,
  2022-01-14T00:00:00.000Z,
  2022-01-15T00:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured time interval for every hour case 1`] = `
Array [
  2022-01-01T09:00:00.000Z,
  2022-01-01T10:00:00.000Z,
  2022-01-01T11:00:00.000Z,
  2022-01-01T12:00:00.000Z,
  2022-01-01T13:00:00.000Z,
  2022-01-01T14:00:00.000Z,
  2022-01-01T15:00:00.000Z,
  2022-01-01T16:00:00.000Z,
  2022-01-01T17:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured time interval for every minute case 1`] = `
Array [
  2022-01-01T12:00:00.000Z,
  2022-01-01T12:01:00.000Z,
  2022-01-01T12:02:00.000Z,
  2022-01-01T12:03:00.000Z,
  2022-01-01T12:04:00.000Z,
  2022-01-01T12:05:00.000Z,
  2022-01-01T12:06:00.000Z,
  2022-01-01T12:07:00.000Z,
  2022-01-01T12:08:00.000Z,
  2022-01-01T12:09:00.000Z,
  2022-01-01T12:10:00.000Z,
  2022-01-01T12:11:00.000Z,
  2022-01-01T12:12:00.000Z,
  2022-01-01T12:13:00.000Z,
  2022-01-01T12:14:00.000Z,
  2022-01-01T12:15:00.000Z,
  2022-01-01T12:16:00.000Z,
  2022-01-01T12:17:00.000Z,
  2022-01-01T12:18:00.000Z,
  2022-01-01T12:19:00.000Z,
  2022-01-01T12:20:00.000Z,
  2022-01-01T12:21:00.000Z,
  2022-01-01T12:22:00.000Z,
  2022-01-01T12:23:00.000Z,
  2022-01-01T12:24:00.000Z,
  2022-01-01T12:25:00.000Z,
  2022-01-01T12:26:00.000Z,
  2022-01-01T12:27:00.000Z,
  2022-01-01T12:28:00.000Z,
  2022-01-01T12:29:00.000Z,
  2022-01-01T12:30:00.000Z,
  2022-01-01T12:31:00.000Z,
  2022-01-01T12:32:00.000Z,
  2022-01-01T12:33:00.000Z,
  2022-01-01T12:34:00.000Z,
  2022-01-01T12:35:00.000Z,
  2022-01-01T12:36:00.000Z,
  2022-01-01T12:37:00.000Z,
  2022-01-01T12:38:00.000Z,
  2022-01-01T12:39:00.000Z,
  2022-01-01T12:40:00.000Z,
  2022-01-01T12:41:00.000Z,
  2022-01-01T12:42:00.000Z,
  2022-01-01T12:43:00.000Z,
  2022-01-01T12:44:00.000Z,
  2022-01-01T12:45:00.000Z,
  2022-01-01T12:46:00.000Z,
  2022-01-01T12:47:00.000Z,
  2022-01-01T12:48:00.000Z,
  2022-01-01T12:49:00.000Z,
  2022-01-01T12:50:00.000Z,
  2022-01-01T12:51:00.000Z,
  2022-01-01T12:52:00.000Z,
  2022-01-01T12:53:00.000Z,
  2022-01-01T12:54:00.000Z,
  2022-01-01T12:55:00.000Z,
  2022-01-01T12:56:00.000Z,
  2022-01-01T12:57:00.000Z,
  2022-01-01T12:58:00.000Z,
  2022-01-01T12:59:00.000Z,
  2022-01-01T13:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured time interval for every month case 1`] = `
Array [
  2022-01-01T00:00:00.000Z,
  2022-02-01T00:00:00.000Z,
  2022-03-01T00:00:00.000Z,
  2022-03-31T23:00:00.000Z,
  2022-04-30T23:00:00.000Z,
  2022-05-31T23:00:00.000Z,
  2022-06-30T23:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured time interval for every two months case 1`] = `
Array [
  2022-01-01T00:00:00.000Z,
  2022-03-01T00:00:00.000Z,
  2022-04-30T23:00:00.000Z,
  2022-06-30T23:00:00.000Z,
  2022-08-31T23:00:00.000Z,
  2022-11-01T00:00:00.000Z,
  2023-01-01T00:00:00.000Z,
]
`;

exports[`TimeScale should create ticks with configured time interval for every year case 1`] = `
Array [
  2021-01-01T00:00:00.000Z,
  2022-01-01T00:00:00.000Z,
  2023-01-01T00:00:00.000Z,
]
`;
